### PR TITLE
Hazeltext - Backend

### PR DESCRIPTION
**Pull request by nj-wilson**
Adds Hazeltext for parsing strings into hazel ASTs.
Reworks a few internals to accommodate this, mainly splitting
out memoization, and moving hazelweb/layout to hazelcore/layout.

Rewrites Examples to be strings ran through the parser.

### CODE REVIEW

File: src/hazelcore/layout/UHAnnot.re [hunk 1]
```diff
@@ -28,6 +28,7 @@ type t =
   | HoleLabel({len: int})
   | Token(token_data)
   | UserNewline
+  | ExternalLineBreak
```
Comment (cyrus-) [inline src/hazelcore/layout/UHAnnot.re]: ExpLineBreak instead of ExternalLineBreak?

File: src/hazeltext/lex.mll [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  ['0'-'9'] ['0'-'9' '_']* ('.' ['0'-'9' '_']*)?
+  (['e' 'E'] ['+' '-']? ['0'-'9'] ['0'-'9' '_'])?
+let float_hex =
+  '0' ['x' 'X'] ['0'-'9' 'A'-'F' 'a'-'f'] ['0'-'9' 'A'-'F' 'a'-'f' '_']*
+  ('.' ['0'-'9' 'A'-'F' 'a'-'f' '_']*)?
+  (['p' 'P'] ['+' '-']? ['0'-'9'] ['0'-'9' '_']*)?
+let float_lit = float | float_hex
+
+let whitespace = [' ' '\160' '\194']
+let newline = ['\n']
```
Comment (cyrus-) [inline src/hazeltext/lex.mll]: windows newlines?

**Review by cyrus- — CHANGES_REQUESTED** [review]

**Review by cyrus- — APPROVED** [review]
Looks great, thanks!

**Comment by dedbox** [comment]
Looks good!

Moving all those files into hazelcore in months-long feature branches is going to suck, but adding type aliases was nearly painless. The grammar is nice and clean.

**Review by dedbox — APPROVED** [review]
Looks good!

Moving all those files into hazelcore in months-long feature branches is going to suck, but adding type aliases was nearly painless. The grammar is nice and clean.

**Pull request merged**